src/Forms/FormField.php

```diff
@@ -444,16 +448,41 @@ public function getInputType()
     }
 
     /**
-     * Returns the field value.
+     * Returns the internal value of the field
+     */
+    public function getValue(): mixed
```

Comment:
We need them all?

getValue() = Get the interval value of the field (this->value), which should be as "unmodified" as possible. Required for the meet FieldValidationInterface
getValueForValidation() = Defaults to $this->getValue(). Something need modify this for some FieldValidators, such as CompositeFieldValidator() which validates child fields rather than the composite field itself
dataValue() = Get a value the for insertion in the database. Will do some casting. If were were to get rid of one of them, I'd say it should be this. But definitely not in this PR.
Value() = Get a value for use in templates. May modify the value quite a bit

Reply:
Yeah, c/d sounds good.
In my mind the name you've suggested there is functionally identical so I don't care if it gets renamed to that or not. This is a field, so all of its validation is "field validation" even if it's not all using a FieldValidator.

That still doesn't solve the problem that we're going from two methods to four in this PR. Most developers will be able to rule out `getValueForValidation()` for their use case, but this PR still makes the decision go from "do I use `value()` or `dataValue()` to now `do I use `value()` or `getValue()` or `dataValue()`?


`getValue()` and `Value()` are so identically named that I would expect them to always return the same value, with one being a legacy alias for the other. 

> `Value()` = Get a value for use in templates. May modify the value quite a bit

I assume this is used via `getDefaultAttributes()` and `getSchemaStateDefaults()`, not directly via `$Value` in templates? If that's the case, it would be sensible to either rename it to something like `getValueForHtmlAttributes()` or else `transformValueForDataAttribute($value)` or something like that.
That makes it immediately clear that if I'm trying to just _get the value from the field in my PHP code_, I don't want to call that one.

That still leaves us with `getValue()` vs `dataValue()`, but at least that confusion already existing before this PR so I'd be okay with leaving it at that point.

Reply:
> I assume this is used via getDefaultAttributes() and getSchemaStateDefaults(), not directly via $Value in templates?

I'm not sure where or how it's exactly used, my best understanding is that it prepares the value for display to a human. Here's a few examples of how it's implemented in subclasses

```php
class DateField extends TextField
// ..
    public function Value()
    {
        return $this->internalToFrontend($this->value);
    }
```

```php
class SiteTreeURLSegmentField extends TextField
// ...
    public function Value()
    {
        return rawurldecode($this->value ?? '');
    }
```

```php
class ReadonlyField
// ...
    public function Value()
    {
        // Get raw value
        $value = $this->dataValue();
        if ($value) {
            return $value;
        }

        // "none" text
        $label = _t('SilverStripe\\Forms\\FormField.NONE', 'none');
        return "<i>('{$label}')</i>";
    }
```

We can't use getValue() as an alias for Value(), as Value() isn't simply returning $this->value

Feels like the best we can do here is to have better PHP doc for both Value() and getValue()

Reply:
> We can't use getValue() as an alias for Value()

I'm not suggesting we do that at all - I'm saying that developers will expect that because they're basically the same named method. The name of the methods is what's problematic here.

I'm suggesting we rename `Value()` to something more specific so that developers don't have any confusion about what the difference between `getValue()` and `Value()` is. PHPDoc will help. Well named methods is better.

> I'm not sure where or how it's exactly used

Please find out where and how it's used, and then give it a name that is appropriate for that usage.
That will make it easier both to know when to call it, and to know what to implement inside that method when implementing new form fields.

Reply:
Keeping things as they are is not an option IMO. We can't add a `getValue()` and not rename `Value()` because the two methods are named too-similarly and it will cause unnecessary confusion.

Reply:
I'll split off as new card, the scope of work is too large (~50 files to change) and will just muddy up this PR. Also a high risk of breaking stuff

Reply:
This PR adds the confusion, so this PR should clear it up, unless you can promise me the split off card is the very next thing you work on.

Reply:
Yes happy to work on that after this is merged